shared/src/main/scala/fetch.scala

```diff
@@ -17,207 +17,467 @@
+import cats.syntax.all._
+import cats.data.NonEmptyList
 
-trait FetchException extends Throwable with Product with Serializable {
-  def env: Env
-}
-case class NotFound(env: Env, request: FetchOne[_, _]) extends FetchException
-case class MissingIdentities(env: Env, missing: Map[DataSourceName, List[Any]])
-    extends FetchException
-case class UnhandledException(env: Env, err: Throwable) extends FetchException
 
-/** Requests in Fetch Free monad. */
-sealed trait FetchRequest extends Product with Serializable
+object `package` {
+  // Fetch queries
+  sealed trait FetchRequest extends Product with Serializable
 
-sealed trait FetchQuery[I, A] extends FetchRequest {
-  def dataSource: DataSource[I, A]
-  def identities: NonEmptyList[I]
-}
+  // A query to a remote data source
+  sealed trait FetchQuery extends FetchRequest {
+    def dataSource[I, A]: DataSource[I, A]
+    def identities[I]: NonEmptyList[I]
+  }
+  case class FetchOne[I, A](id: I, ds: DataSource[I, A]) extends FetchQuery {
+    override def identities[I]: NonEmptyList[I] = NonEmptyList(id.asInstanceOf[I], List.empty[I])
+    override def dataSource[I, A]: DataSource[I, A] = ds.asInstanceOf[DataSource[I, A]]
+  }
+  case class Batch[I, A](ids: NonEmptyList[I], ds: DataSource[I, A]) extends FetchQuery {
+    override def identities[I]: NonEmptyList[I] = ids.toNonEmptyList.asInstanceOf[NonEmptyList[I]]
+    override def dataSource[I, A]: DataSource[I, A] = ds.asInstanceOf[DataSource[I, A]]
+  }
 
-/**
- * Primitive operations in the Fetch Free monad.
- */
-sealed abstract class FetchOp[A] extends Product with Serializable
+  // Fetch result states
+  sealed trait FetchStatus
+  case class FetchDone[A](result: A) extends FetchStatus
+  case class FetchMissing() extends FetchStatus
 
-final case class FetchOne[I, A](id: I, ds: DataSource[I, A])
-    extends FetchOp[A]
-    with FetchQuery[I, A] {
-  override def dataSource: DataSource[I, A] = ds
-  override def identities: NonEmptyList[I]  = NonEmptyList.one(id)
-}
+  // Fetch errors
+  sealed trait FetchException extends Throwable with NoStackTrace {
+    def environment: Env
+  }
+  case class MissingIdentity[I](i: I, request: FetchQuery, environment: Env) extends FetchException
+  case class UnhandledException(e: Throwable, environment: Env) extends FetchException
 
-final case class FetchMany[I, A](ids: NonEmptyList[I], ds: DataSource[I, A])
-    extends FetchOp[List[A]]
-    with FetchQuery[I, A] {
-  override def dataSource: DataSource[I, A] = ds
-  override def identities: NonEmptyList[I]  = ids
-}
+  // In-progress request
+  case class BlockedRequest(request: FetchRequest, result: FetchStatus => IO[Unit])
 
-final case class Concurrent(queries: NonEmptyList[FetchQuery[Any, Any]])
-    extends FetchOp[InMemoryCache]
-    with FetchRequest
+  /* Combines the identities of two `FetchQuery` to the same data source. */
+  private def combineIdentities[I](x: FetchQuery, y: FetchQuery): NonEmptyList[I] = {
```

Comment:
This seems like a dangerous method due to the reason stated above. We know nothing about the underlying `Datastore`, including their key and value types, yet we specify below that their identity types will be the same. If the caller does not enforce this guarantee, won't this blow up?